pkg/capacity/capacity.go

```diff
@@ -0,0 +1,627 @@
+		scInformer:       scInformer,
+		cInformer:        cInformer,
+		pollPeriod:       pollPeriod,
+		immediateBinding: immediateBinding,
+		capacities:       map[workItem]*storagev1alpha1.CSIStorageCapacity{},
+	}
+
+	// Now register for changes. Depending on the implementation of the informers,
+	// this may already invoke callbacks.
+	handler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			sc, ok := obj.(*storagev1.StorageClass)
+			if !ok {
+				klog.Errorf("added object: expected StorageClass, got %T -> ignoring it", obj)
+				return
+			}
+			c.onSCAddOrUpdate(sc)
+		},
+		UpdateFunc: func(_ interface{}, newObj interface{}) {
+			sc, ok := newObj.(*storagev1.StorageClass)
+			if !ok {
+				klog.Errorf("updated object: expected StorageClass, got %T -> ignoring it", newObj)
+				return
+			}
+			c.onSCAddOrUpdate(sc)
+		},
+		DeleteFunc: func(obj interface{}) {
+			// Beware of "xxx deleted" events
+			if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+				obj = unknown.Obj
+			}
+			sc, ok := obj.(*storagev1.StorageClass)
+			if !ok {
+				klog.Errorf("deleted object: expected StorageClass, got %T -> ignoring it", obj)
+				return
+			}
+			c.onSCDelete(sc)
+		},
+	}
+	c.scInformer.Informer().AddEventHandler(handler)
+	c.topologyInformer.AddCallback(c.onTopologyChanges)
+
+	// We don't want the callbacks yet, but need to ensure that
+	// the informer controller is instantiated before the caller
+	// starts the factory.
+	cInformer.Informer()
+
+	return c
+}
+
+// Run is a main Controller handler
+func (c *Controller) Run(ctx context.Context, threadiness int) {
+	klog.Info("Starting Capacity Controller")
+	defer c.queue.ShutDown()
+	go c.scInformer.Informer().Run(ctx.Done())
+	go c.topologyInformer.Run(ctx)
+
+	c.prepare(ctx)
+	for i := 0; i < threadiness; i++ {
+		go wait.UntilWithContext(ctx, func(ctx context.Context) {
+			c.runWorker(ctx)
+		}, time.Second)
+	}
+
+	go wait.UntilWithContext(ctx, func(ctx context.Context) { c.pollCapacities() }, c.pollPeriod)
+
+	klog.Info("Started Capacity Controller")
+	<-ctx.Done()
+	klog.Info("Shutting down Capacity Controller")
+}
+
+func (c *Controller) prepare(ctx context.Context) {
+	// Wait for topology and storage class informer sync. Once we have that,
+	// we know which CSIStorageCapacity objects we need.
+	if !cache.WaitForCacheSync(ctx.Done(), c.topologyInformer.HasSynced, c.scInformer.Informer().HasSynced, c.cInformer.Informer().HasSynced) {
+		return
+	}
+
+	// The caches are fully populated now, but the event handlers
+	// may or may not have been invoked yet. To be sure that we
+	// have all data, we need to list all resources. Here we list
+	// topology segments, onTopologyChanges lists the classes.
+	c.onTopologyChanges(c.topologyInformer.List(), nil)
+
+	klog.V(3).Infof("Initial number of topology segments %d, storage classes %d, potential CSIStorageCapacity objects %d",
+		len(c.topologyInformer.List()),
+		func() int {
```

Comment:
Can we store this somewhere else? It would be nice to not have to do a List just for a log message.

Reply:
It's in a function because I didn't want to introduce variables that are only used inside the log call. I've changed it.

Both `c.topologyInformer` and `c.scInformer` implement `List` as a local operation and this log message is only printed once, therefore I am not worried about the performance.